src/libpriv/rpmostree-core.cxx

```diff
@@ -1997,14 +2001,50 @@ rpmostree_context_prepare (RpmOstreeContext *self,
     }
 
   /* First, handle packages to remove */
-  g_autoptr(GPtrArray) removed_pkgnames = g_ptr_array_new ();
+  g_autoptr(GPtrArray) removed_pkgnames = g_ptr_array_new_with_free_func (g_free);
   for (auto &pkgname_v: packages_override_remove)
     {
       const char *pkgname = pkgname_v.c_str();
       if (!dnf_context_remove (dnfctx, pkgname, error))
         return FALSE;
 
-      g_ptr_array_add (removed_pkgnames, (gpointer)pkgname);
+      g_ptr_array_add (removed_pkgnames, g_strdup(pkgname));
+    }
+
+  // Then, repo packages to remove and replace
+  g_autoptr(DnfPackageSet) pinned_pkgs = dnf_packageset_new (sack);
+  Map *pinned_pkgs_map = dnf_packageset_get_map (pinned_pkgs);
+  auto override_packages_query = self->treefile_rs->get_override_replace_query();
+  for (auto & repo_pkg : override_packages_query)
+    {
+      auto repo = std::string(repo_pkg.get_repo());
+      auto pkgs = repo_pkg.get_packages();
+      for (auto & pkg_r : pkgs)
+        {
+          // We need to copy here because we only have an immutable reference to Rust
+          auto pkg = std::string(pkg_r);
+          const char *pkgname = pkg.c_str();
+          if (!dnf_context_remove (dnfctx, pkgname, error))
+            return FALSE;
+          // Then copy here *again*; TODO: switch this to a vec which owns the string
+          char *copied = g_strdup (pkgname);
+          g_ptr_array_add (replaced_pkgnames, copied);
+
+          g_auto(HySubject) subject = hy_subject_create(pkg.c_str());
+          /* this is basically like a slightly customized dnf_context_install() */
+          HyNevra nevra = NULL;
+          hy_autoquery HyQuery query =
+            hy_subject_get_best_solution(subject, sack, NULL, &nevra, FALSE, TRUE, TRUE, TRUE, FALSE);
```

Comment:
Hmm, instead of this, I think we should be querying strictly by package name. `repo-packages` does it this way because it wants to provide the same semantics as `packages` (which supports e.g. provides and filepaths). But for overrides I think we can be stricter and require the exact package name to override from the base.

So then from the query we'll be getting specific NEVRAs we can add to `replaced_nevras` as usual.